### PR TITLE
ViewModel architecture component reference app

### PR DESCRIPTION
**Pull request by patrickhammond**

### CODE REVIEW

**Comment by patrickhammond** [comment]
Any feedback @atoennis or @stuartsoft ?

**Comment by atoennis** [comment]
I know `BaseActivity.kt` wasn't added to this PR but base classes tend to scare me 😱 .  I think they can easily grow to have a lot of disjointed functionality.  The new functionality that was added was to get the view model (also added to `BaseFragment.kt`).  Do you think the convenience is worth the cost of adding to the inheritance chain?

If we think it's good to keep, how about renaming it to convey a more specific intent.  This may not be a good name but something like `DataBindingActivity.kt` would encourage disjointed functionality from being added.

**Comment by patrickhammond** [comment]
Several times in the last few months I've had to fix Android specific
version bugs because something wasn't inheriting from AppCompatActivity or
the support Fragment implementation (resolved by simply changing the
import). My goal with establishing those base classes is to make it harder
for those types of issues to sneak in. I tend to agree that base classes
can become disjointed...so one step back two steps forward?

**Comment by atoennis** [comment]
There are a few things I discussed with Patrick that needs to be fixed:

1. Data binding layout inflator.  Should we go with A or B?
  A. `CommitItemBinding.inflate(layoutInflater, parent, false)`
  B. `DataBindingUtil.inflate(inflater, R.layout.fragment_main, container, false)`
2. Bug with the `initialized` property being persisted in `State`
  A. Should be persisted on the ViewModel.
3. `BaseViewModel` has code copied and pasted from `BaseObservable`.  Instead, pass an instance of `BaseObservable` into `BaseViewModel` and pass into the `BaseObservable` (composition)

**Comment by patrickhammond** [comment]
1 and 2 are out there now. When I try to do 3, databinding breaks and I'm
not sure why yet.

**Comment by patrickhammond** [comment]
It seems BaseObservable isn't composition friendly. See: https://android.jlelse.eu/make-your-view-models-great-again-ead9ee98f4f2

**Comment by atoennis** [comment]
I read through the article you linked but didn't quite see why BaseObservable isn't composition friendly.  I'll play with it this afternoon if I get time, I'm guessing I'll be able to see why it isn't easier that way.

File: app/src/main/java/com/mycompany/myapp/util/RxUtils.kt [hunk 1]
```diff
@@ -1,12 +1,12 @@
 package com.mycompany.myapp.util
 
-import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.Observable
+import io.reactivex.functions.BiFunction
+import java.util.concurrent.TimeUnit
 
 object RxUtils {
-    fun getSafeCompositeDisposable(disposable: CompositeDisposable?): CompositeDisposable {
```
Comment (atoennis) [inline app/src/main/java/com/mycompany/myapp/util/RxUtils.kt]: Was this previously unused?

File: app/src/main/java/com/mycompany/myapp/util/RxUtils.kt [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
-        return when {
-            disposable == null || disposable.isDisposed -> CompositeDisposable()
-            else -> disposable
-        }
+    fun <T> delayAtLeast(observable: Observable<T>, delayMs: Long): Observable<T> {
```
Comment (atoennis) [inline app/src/main/java/com/mycompany/myapp/util/RxUtils.kt]: Good refactoring, this is a much better home 👍

File: app/src/main/java/com/mycompany/myapp/ui/main/MainViewModel.kt [hunk 3]
```diff
@@ -75,35 +77,28 @@ class MainViewModel @Inject constructor(
...
+                        .doFinally({ loadingCommits = false })
+                        .subscribeOn(ioScheduler)
+                        .observeOn(mainScheduler)
+                        .subscribe({ commits = it }, { handleError(it) }))
     }
 
-    private fun <T> delayAtLeast(observable: Observable<T>, delayMs: Long): Observable<T> {
-        val timer = Observable.timer(delayMs, TimeUnit.MILLISECONDS)
-        return Observable.combineLatest<Long, T, T>(timer, observable, BiFunction { _, response -> response })
+    private fun loadCommits(username: String, repository: String): Observable<List<Commit>> {
```
Comment (atoennis) [inline app/src/main/java/com/mycompany/myapp/ui/main/MainViewModel.kt]: I like what you did with this function, much cleaner!

File: app/src/main/java/com/mycompany/myapp/ui/main/MainViewModel.kt [hunk 4]
```diff
@@ -75,35 +77,28 @@ class MainViewModel @Inject constructor(
...
-
-    private fun loadCommits(request: LoadCommitsRequest): Disposable {
-        return delayAtLeast(gitHubInteractor.loadCommits(request), loadingDelayMs)
-                .flatMap<Commit> { response -> Observable.fromIterable<Commit>(response.commits) }
-                .map(this::toCommitView)
-                .toList()
-                .subscribeOn(ioScheduler)
-                .observeOn(mainScheduler)
-                .subscribe({ commits = it }, { handleError(it) })
+        loadingCommits = true
```
Comment (atoennis) [inline app/src/main/java/com/mycompany/myapp/ui/main/MainViewModel.kt]: Moving the mapping of commits to `CommitViews` makes sense in this method 👍

File: app/src/main/java/com/mycompany/myapp/util/RxUtils.kt (hunk 1 above)
Comment (patrickhammond) [reply app/src/main/java/com/mycompany/myapp/util/RxUtils.kt]: Yes

File: app/src/main/java/com/mycompany/myapp/ui/main/MainViewModel.kt [hunk 5]
```diff
@@ -43,7 +42,7 @@ class MainViewModel @Inject constructor(
         fetchCommits()
     }
 
-    private var commits: Commits = Commits.Result(emptyList())
+    internal var commits: Commits = Commits.Result(emptyList())
```
Comment (atoennis) [inline app/src/main/java/com/mycompany/myapp/ui/main/MainViewModel.kt]: @patrickhammond Had to change the scope of this guy for testing purposes.  I needed a handle to `commits` while unit testing the `fetchCommits()` method.  Not too happy about it but curious what you think.

Comment (patrickhammond) [reply app/src/main/java/com/mycompany/myapp/ui/main/MainViewModel.kt]: That seems 100% legit to me. I've started defaulting to `internal` anywhere I'd have package scope in Java. For better or for worse. @weverb2 any guidance or opinions you have on best practices around Kotlin access modifiers wrt to supporting testing?

Comment (weverb2) [reply app/src/main/java/com/mycompany/myapp/ui/main/MainViewModel.kt]: @atoennis the only thing I would have to say is maybe mark it `@VisibleForTesting`?

Comment (atoennis) [reply app/src/main/java/com/mycompany/myapp/ui/main/MainViewModel.kt]: Done!

**Pull request merged**